rpc/dial.go

```diff
@@ -107,77 +118,118 @@ func dial(
+					logger.Debug("using signaling's external auth as auth material")
+					dOpts.authMaterial = dOpts.webrtcOpts.SignalingExternalAuthAuthMaterial
+					dOpts.creds = Credentials{}
+				}
 			}
-			fixupWebRTCOptions(dOpts, target, port)
-
-			// When connecting to an external signaler for WebRTC, we assume we can use the external auth's material.
-			// This path is also called by an mdns direct connection and ignores that case.
-			// This will skip all Authenticate/AuthenticateTo calls for the signaler.
-			if !dOpts.usingMDNS && dOpts.authMaterial == "" && dOpts.webrtcOpts.SignalingExternalAuthAuthMaterial != "" {
-				logger.Debug("using signaling's external auth as auth material")
-				dOpts.authMaterial = dOpts.webrtcOpts.SignalingExternalAuthAuthMaterial
-				dOpts.creds = Credentials{}
+
+			if dOpts.debug {
+				logger.Debugw(
+					"trying WebRTC",
+					"signaling_server", dOpts.webrtcOpts.SignalingServerAddress,
+					"host", originalAddress,
+				)
 			}
-		}
 
-		if dOpts.debug {
-			logger.Debugw(
-				"trying WebRTC",
-				"signaling_server", dOpts.webrtcOpts.SignalingServerAddress,
-				"host", originalAddress,
-			)
-		}
+			conn, cached, err := dialFunc(
+				ctx,
+				"webrtc",
+				fmt.Sprintf("%s->%s", dOpts.webrtcOpts.SignalingServerAddress, originalAddress),
+				buildKeyExtra(dOpts),
+				func() (ClientConn, error) {
+					return dialWebRTC(
+						ctx,
+						dOpts.webrtcOpts.SignalingServerAddress,
+						originalAddress,
+						dOpts,
+						logger,
+					)
+				})
+			if err == nil {
+				if !cached {
+					logger.Debug("connected via WebRTC")
+				} else if dOpts.debug {
+					logger.Debug("connected via WebRTC (cached)")
+				}
+				connCh <- dialSuccess{conn, cached}
+				return
+			}
+			if !errors.Is(err, ErrNoWebRTCSignaler) {
+				errCh <- dialError{err, true}
+				return
+			}
+			if ctx.Err() != nil {
+				errCh <- dialError{err, true}
+				return
+			}
+			errCh <- dialError{err, false}
+		}()
+	}
 
-		conn, cached, err := dialFunc(
-			ctx,
-			"webrtc",
-			fmt.Sprintf("%s->%s", dOpts.webrtcOpts.SignalingServerAddress, originalAddress),
-			buildKeyExtra(dOpts),
-			func() (ClientConn, error) {
-				return dialWebRTC(
-					ctx,
-					dOpts.webrtcOpts.SignalingServerAddress,
-					originalAddress,
-					dOpts,
-					logger,
-				)
-			})
-		if err == nil {
-			if !cached {
-				logger.Debug("connected via WebRTC")
-			} else if dOpts.debug {
-				logger.Debug("connected via WebRTC (cached)")
+	var (
+		dialErrors []error
+		exit       bool
+	)
+	for len(dialErrors) < dialAttempts {
+		select {
+		case success := <-connCh:
+			return success.conn, success.cached, nil
+		case err := <-errCh:
+			dialErrors = append(dialErrors, err.err)
+			if err.exit {
+				exit = true
 			}
-			return conn, cached, nil
-		}
-		if !errors.Is(err, ErrNoWebRTCSignaler) {
-			return nil, false, err
-		}
-		if ctx.Err() != nil {
-			return nil, false, ctx.Err()
```

Comment:
It's not clear to me why we exit early on these particular errors without making an attempt to dial directly through gRPC.

Reply:
the context one makes sense, and tho the one above I'm not sure about - maybe we know for sure that if there is a webrtc signaler and we can't connect there's no chance of the connection succeeding?

Reply:
in any case, I made [a follow-up ticket](https://viam.atlassian.net/browse/RSDK-6493) to investigate or possibly refactor this